lib/cea/mp4_cea_parser.js

```diff
@@ -0,0 +1,229 @@
+   * and then parses the MDAT boxes for CEA-708 packets using this information.
+   * CEA-708 packets are returned in the callback.
+   * @override
+   */
+  parse(data, on708Data) {
+    const Mp4Parser = shaka.util.Mp4Parser;
+
+    // Fields that are found in MOOF boxes
+    let defaultSampleDuration = this.defaultSampleDuration_;
+    let defaultSampleSize = this.defaultSampleSize_;
+    let sampleData = [];
+    let baseMediaDecodeTime = null;
+    let timescale = shaka.cea.CeaParser.DEFAULT_TIMESCALE_VALUE;
+
+    new Mp4Parser()
+        .box('moof', Mp4Parser.children)
+        .box('traf', Mp4Parser.children)
+        .fullBox('trun', (box) => {
+          goog.asserts.assert(
+              box.version != null && box.flags!=null,
+              'TRUN is a full box and should have a valid version & flags.');
+
+          const parsedTRUN = shaka.util.Mp4BoxParsers.parseTRUN(
+              box.reader, box.version, box.flags);
+
+          sampleData = parsedTRUN.sampleData;
+        })
+
+        .fullBox('tfhd', (box) => {
+          goog.asserts.assert(
+              box.flags != null,
+              'TFHD is a full box and should have valid flags.');
+
+          const parsedTFHD = shaka.util.Mp4BoxParsers.parseTFHD(
+              box.reader, box.flags);
+
+          // If specified, defaultSampleDuration and defaultSampleSize
+          // override the ones specified in the TREX box
+          defaultSampleDuration = parsedTFHD.defaultSampleDuration
+            || this.defaultSampleDuration_;
+
+          defaultSampleSize = parsedTFHD.defaultSampleSize
+            || this.defaultSampleSize_;
+
+          const trackId = parsedTFHD.trackId;
+
+          // Get the timescale from the track Id
+          if (this.trackIdToTimescale_[trackId]) {
+            timescale = this.trackIdToTimescale_[trackId];
+          }
+        })
+
+        .fullBox('tfdt', (box) => {
+          goog.asserts.assert(
+              box.version != null,
+              'TFDT is a full box and should have a valid version.');
+
+          const parsedTFDT = shaka.util.Mp4BoxParsers.parseTFDT(
+              box.reader, box.version);
+
+          baseMediaDecodeTime = parsedTFDT.baseMediaDecodeTime;
+        })
+        .box('mdat', (box) => {
+          this.parseMdat_(box.reader, baseMediaDecodeTime, timescale,
+              defaultSampleDuration, defaultSampleSize, sampleData, on708Data);
+        })
+        .parse(data, /* partialOkay= */ false);
+  }
+
+  /**
+   * Parse MDAT box.
+   * @param {!shaka.util.DataViewReader} reader
+   * @param {?number} time
+   * @param {!number} timescale
+   * @param {!number} defaultSampleDuration
+   * @param {!number} defaultSampleSize
+   * @param {!Array<shaka.util.ParsedTRUNSample>} sampleData
+   * @param {function(!BufferSource, number)} on708Data
+   * @private
+   */
+  parseMdat_(reader, time, timescale, defaultSampleDuration,
+      defaultSampleSize, sampleData, on708Data) {
+    if (time === null) {
```

Comment:
Would it be better if we check the time before line 172?

Reply:
Yes, I agree that makes more sense to do the null check there.